File: data/trainings-data/src/main/java/app/fit/fitndflow/data/trainings/repositoryImpl/TrainingRepositoryImpl.java
package app.fit.fitndflow.data.trainings.repositoryImpl;

import android.content.Context;

import app.fit.fitndflow.data.common.datasource.TrainingLocalDataSource;
import app.fit.fitndflow.data.common.dto.CategoryDto;
import app.fit.fitndflow.data.common.dto.ExerciseDto;
import app.fit.fitndflow.data.common.dto.SerieDto;
import app.fit.fitndflow.data.common.mapper.CategoryModelMapperKt;
import app.fit.fitndflow.data.common.mapper.ExerciseModelMapperKt;

import java.util.ArrayList;
import java.util.List;

import app.fit.fitndflow.data.common.model.ExcepcionApi;
import app.fit.fitndflow.data.trainings.datasource.remote.TrainingRemoteDataSource;
import app.fit.fitndflow.data.trainings.dto.AddSerieRequestDto;
import app.fit.fitndflow.data.trainings.dto.SerieForAddSerieRequestDto;
import app.fit.fitndflow.data.trainings.model.TrainingsApiInterface;

import com.fit.fitndflow.app.domain.common.models.CategoryModel;
import com.fit.fitndflow.app.domain.common.models.ExerciseModel;
import com.fit.fitndflow.app.domain.common.models.SerieModel;
import com.fit.fitndflow.app.domain.trainings.repository.TrainingRepository;

import retrofit2.Response;

public class TrainingRepositoryImpl implements TrainingRepository {
    private TrainingLocalDataSource trainingLocalDataSource;
    private Context mContext;
    private TrainingRemoteDataSource trainingRemoteDataSource;

    public TrainingRepositoryImpl(Context context, TrainingRemoteDataSource trainingRemoteDataSource, TrainingLocalDataSource trainingLocalDataSource) {
        this.mContext = context;
        this.trainingRemoteDataSource = trainingRemoteDataSource;
        this.trainingLocalDataSource = trainingLocalDataSource;
    }

    @Override
    public List<SerieModel> getSerieListOfExerciseAdded(int exerciseId) throws Exception {
        try {
            List<CategoryModel> categoryList = trainingLocalDataSource.getTrainingsByDate(trainingLocalDataSource.getCurrentDate());
            if (categoryList != null) {
                for (int i = 0; i < categoryList.size(); i++) {
                    CategoryModel category = categoryList.get(i);
                    List<ExerciseModel> exerciseList = category.getExerciseList();
                    for (int j = 0; j < exerciseList.size(); j++) {
                        ExerciseModel exercise = exerciseList.get(j);
                        if (exercise.getId() == exerciseId) {
                            return exercise.getSerieList();
                        }
                    }
                }
            }
            return new ArrayList<>();
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
    }

    @Override
    public ExerciseModel addNewSerie(int reps, double weight, int exerciseId) throws Exception {
        ExerciseModel exerciseResponse;
        ExerciseDto response;
        try {
            response = trainingRemoteDataSource.addNewSerie(reps, weight, exerciseId, trainingLocalDataSource.getCurrentDate());
            exerciseResponse = ExerciseModelMapperKt.toModel(response);
            trainingLocalDataSource.cleanCache();
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
        return exerciseResponse;
    }

    @Override
    public ExerciseModel modifySerie(int serieId, int reps, double weight) throws Exception {
        ExerciseModel exerciseResponse;
        ExerciseDto response;
        try {
            response = trainingRemoteDataSource.modifySerie(serieId, reps, weight);
            exerciseResponse = ExerciseModelMapperKt.toModel(response);
            trainingLocalDataSource.cleanCache();

        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
        return exerciseResponse;
    }

    @Override
    public ExerciseModel deleteSerie(int serieId) throws Exception {
        ExerciseModel exerciseResponse;
        ExerciseDto response;
        try {
            response = trainingRemoteDataSource.deleteSerie(serieId);
            exerciseResponse = ExerciseModelMapperKt.toModel(response);
            trainingLocalDataSource.cleanCache();
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
        return exerciseResponse;
    }

    @Override
    public List<CategoryModel> getTrainingListAndUpdateCache(String date) throws Exception {
        trainingLocalDataSource.setCurrentDate(date);
        List<CategoryDto> response;
        if (trainingLocalDataSource.getTrainingsByDate(date) == null) {
            try {
                response = trainingRemoteDataSource.getTrainingListAndUpdateCache(date);
                trainingLocalDataSource.replaceAllDataFromTrainingCache(date, CategoryModelMapperKt.toModel(response));
            } catch (Exception e) {
                e.printStackTrace();
                throw new Exception(e);
            }
        }
        return trainingLocalDataSource.getTrainingsByDate(date);
    }

    @Override
    public List<CategoryModel> updateCurrentTrainingListCache() throws Exception {
        return getTrainingListAndUpdateCache(trainingLocalDataSource.getCurrentDate());
    }
}

File: data/trainings-data/src/main/java/app/fit/fitndflow/data/trainings/repositoryImpl/TrainingRepositoryImpl.kt
package app.fit.fitndflow.data.trainings.repositoryImpl

import android.content.Context
import app.fit.fitndflow.data.common.database.dao.ExerciseDao
import app.fit.fitndflow.data.common.database.dao.SerieDao
import app.fit.fitndflow.data.common.database.dao.TrainingDao
import app.fit.fitndflow.data.common.database.entities.SerieEntity
import app.fit.fitndflow.data.common.database.entities.filterTrainingByDate
import app.fit.fitndflow.data.common.database.mapper.toModel
import app.fit.fitndflow.data.common.datasource.local.SharedPrefsLocalDataSource
import app.fit.fitndflow.data.common.datasource.local.TrainingCacheLocalDataSource
import app.fit.fitndflow.data.common.mapper.CategoryModelMapperKt.Companion.toModel
import app.fit.fitndflow.data.common.mapper.ExerciseModelMapperKt.Companion.toModel
import app.fit.fitndflow.data.trainings.datasource.remote.TrainingRemoteDataSource
import com.fit.fitndflow.app.domain.common.models.CategoryModel
import com.fit.fitndflow.app.domain.common.models.ExerciseModel
import com.fit.fitndflow.app.domain.common.models.SerieModel
import com.fit.fitndflow.app.domain.trainings.repository.TrainingRepository

class TrainingRepositoryImpl(
    private val mContext: Context,
    private val trainingRemoteDataSource: TrainingRemoteDataSource,
    private val trainingCacheLocalDataSource: TrainingCacheLocalDataSource,
    private val serieDao: SerieDao,
    private val trainingDao: TrainingDao,
    private val exerciseDao: ExerciseDao,
    private val sharedPrefsLocalDataSource: SharedPrefsLocalDataSource
) : TrainingRepository {
    private val isLocalMode
        get() = sharedPrefsLocalDataSource.getIsLocal()

    override suspend fun getSerieListOfExerciseAdded(exerciseId: Int): List<SerieModel> {
        return try {
            val currentDate = trainingCacheLocalDataSource.currentDate ?: return emptyList()
            val categoryList = trainingCacheLocalDataSource.getTrainingsByDate(currentDate)
            categoryList?.forEach { category ->
                category.exerciseList?.forEach { exercise ->
                    if (exercise.id == exerciseId) {
                        return exercise.serieList.toList()
                    }
                }
            }
            emptyList()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
    }

    override suspend fun addNewSerie(reps: Int, weight: Double, exerciseId: Int): ExerciseModel? {
        var response: ExerciseModel? = null
        try {
            trainingCacheLocalDataSource.currentDate?.let { currentDate ->
                if (isLocalMode) {
                    if(getSerieListOfExerciseAdded(exerciseId).isEmpty()) exerciseDao.updateFirstSerie(exerciseId, reps, weight)

                    serieDao.insertSerie(
                        SerieEntity(
                            exerciseId = exerciseId,
                            reps = reps,
                            weight = weight,
                            date = currentDate
                        )
                    )
                    val exercise = exerciseDao.getExercise(exerciseId).firstOrNull()
                    val currentRecord = exercise
                        ?.takeUnless { it.recordHeight == 0.0 && it.recordReps == 0 }
                        ?.let { SerieModel(reps = it.recordReps, kg = it.recordHeight, isRecord = true) }
                    val isRecord = currentRecord == null ||
                        weight > (currentRecord.kg ?: 0.0) ||
                        (weight >= (currentRecord.kg ?: 0.0) && reps > (currentRecord.reps ?: 0))
                    val lastRecord = if (isRecord) {
                        SerieModel(reps = reps, kg = weight, isRecord = true)
                    } else {
                        currentRecord
                    }
                    if (isRecord) { exerciseDao.updateRecord(exerciseId, weight, reps) }
                    response =
                        serieDao.getExerciseWithSeriesByDate(exerciseId, currentDate)
                            .toModel(lastRecord)
                } else {
                    trainingRemoteDataSource.addNewSerie(
                        reps,
                        weight,
                        exerciseId,
                        currentDate
                    )?.let {
                        response = toModel(it)
                    }
                }
            }
            trainingCacheLocalDataSource.cleanCache()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
        return response
    }

    override suspend fun modifySerie(serieId: Int, reps: Int, weight: Double): ExerciseModel? {
        var response: ExerciseModel? = null
        try {
            trainingCacheLocalDataSource.currentDate?.let { currentDate ->
                if (isLocalMode) {
                    val exerciseId = serieDao.getSerie(serieId).exerciseId
                    serieDao.updateSerie(
                        serieId = serieId,
                        reps = reps,
                        weight = weight
                    )
                    response =
                        serieDao.getExerciseWithSeries(exerciseId, currentDate).first().toModel()
                } else {
                    trainingRemoteDataSource.modifySerie(serieId, reps, weight)?.let { response = toModel(it) }
                }
            }
            trainingCacheLocalDataSource.cleanCache()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
        return response
    }

    override suspend fun deleteSerie(serieId: Int): ExerciseModel? {
        var response: ExerciseModel? = null
        try {
            trainingCacheLocalDataSource.currentDate?.let { currentDate ->
                if (isLocalMode) {
                    val exerciseId = serieDao.getSerie(serieId).exerciseId
                    serieDao.deleteSerie(serieId)
                    response = serieDao.getExerciseWithSeries(exerciseId,currentDate).first().toModel()
                } else {
                    trainingRemoteDataSource.deleteSerie(serieId)?.let { response = toModel(it) }
                }
            }
            trainingCacheLocalDataSource.cleanCache()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
        return response
    }

    override suspend fun getTrainingListAndUpdateCache(date: String): List<CategoryModel> {
        trainingCacheLocalDataSource.currentDate = date
        val response: List<CategoryModel>
        if (trainingCacheLocalDataSource.getTrainingsByDate(date) == null) {
            try {
                if (isLocalMode){
                    response = filterTrainingByDate(trainingDao.getFullTraining(), date).map { it.toModel() }
                } else {
                    response = toModel(trainingRemoteDataSource.getTrainingListAndUpdateCache(date))
                }
                    trainingCacheLocalDataSource.replaceAllDataFromTrainingCache(date, response)
            } catch (e: Exception) {
                e.printStackTrace()
                throw Exception(e)
            }
        }
        return trainingCacheLocalDataSource.getTrainingsByDate(date).orEmpty()
    }

    override suspend fun updateCurrentTrainingListCache(): List<CategoryModel> {
        return getTrainingListAndUpdateCache(trainingCacheLocalDataSource.currentDate.orEmpty())
    }
}
